lsif/src/backend.ts

```diff
@@ -0,0 +1,150 @@
+     * dump. Insert these contents into some storage with an encoding that
+     * can be subsequently read by the `createRunner` method.
+     */
+    public async insertDump(input: Readable, repository: string, commit: string): Promise<void> {
+        const outFile = makeFilename(this.storageRoot, repository, commit)
+
+        try {
+            await fs.unlink(outFile)
+        } catch (e) {
+            if (!hasErrorCode(e, 'ENOENT')) {
+                throw e
+            }
+        }
+
+        const { packages, references } = await this.connectionCache.withTransactionalEntityManager(
+            outFile,
+            [DefModel, DocumentModel, MetaModel, RefModel],
+            (entityManager: EntityManager) => importLsif(entityManager, parseLines(readline.createInterface({ input })))
+        )
+
+        // These needs to be done in sequence as SQLite can only have one
+        // write txn at a time without causing the other one to abort with
+        // a weird error.
+        await this.xrepoDatabase.addPackages(repository, commit, packages)
+        await this.xrepoDatabase.addReferences(repository, commit, references)
+    }
+
+    /**
+     * Create a database relevant to the given repository and commit hash. This
+     * assumes that data for this subset of data has already been inserted via
+     * `insertDump` (otherwise this method is expected to throw).
+     */
+    public async createDatabase(repository: string, commit: string): Promise<Database> {
+        const file = makeFilename(this.storageRoot, repository, commit)
+
+        try {
+            await fs.stat(file)
+        } catch (e) {
+            if (hasErrorCode(e, 'ENOENT')) {
+                throw new NoLSIFDataError(repository, commit)
+            }
+
+            throw e
+        }
+
+        return new Database(
+            this.storageRoot,
+            this.xrepoDatabase,
+            this.connectionCache,
+            this.documentCache,
+            repository,
+            commit,
+            file
+        )
+    }
+}
+
+/**
+ * Create the path of the SQLite database file for the given repository and commit.
+ *
+ * @param storageRoot The path where SQLite databases are stored.
+ * @param repository The repository name.
+ * @param commit The repository commit.
+ */
+export function makeFilename(storageRoot: string, repository: string, commit: string): string {
+    return path.join(storageRoot, `${encodeURIComponent(repository)}@${commit}.lsif.db`)
+}
+
+/**
+ * Converts streaming JSON input into an iterable of vertex and edge objects.
+ *
+ * @param lines The stream of raw, uncompressed JSON lines.
+ */
+async function* parseLines(lines: AsyncIterable<string>): AsyncIterable<Vertex | Edge> {
+    let i = 0
+    for await (const line of lines) {
+        try {
+            yield JSON.parse(line) as Vertex | Edge
+        } catch (e) {
+            throw new Error(`Parsing failed for line ${i}: ${e}`)
```

Comment:
Don't include the `e.toString()` inside the error message of another error, it will make the error message very long as it puts the stack trace into the message.
```suggestion
            throw Object.assign(new Error(`Parsing failed for line ${i}`), { error })
```